vllm/model_executor/layers/moe.py

```diff
@@ -0,0 +1,354 @@
+from typing import Tuple
+
+import torch
+
+from torch import nn
+import torch.nn.functional as F
+
+import triton
+import triton.language as tl
+
+from vllm._C import ops
+from vllm.model_executor.layers.linear import ReplicatedLinear
+
+from vllm.model_executor.parallel_utils.parallel_state import (
+    get_tensor_model_parallel_rank)
+from vllm.model_executor.parallel_utils.communication_op import (
+    tensor_model_parallel_all_reduce)
+from vllm.model_executor.utils import set_weight_attrs
+
+
+class MoE(nn.Module):
+    """A expert parallel MOE that shards each expert across all ranks.
```

Comment:
As I commented above, I think this is a tensor-parallel implementation of MoE, rather than the expert parallel one, since we do not split the model along the expert dimension.